AsyncLocalStorage context lost across node:net callbacks

Version: Deno 2.8.2
OS: Ubuntu 24.04.4 LTS

### Description
AsyncLocalStorage context is not preserved across callbacks in the node:net compatibility layer.


### Steps to Reproduce
<!-- Provide a minimal code snippet or link to a repository that causes the issue -->
1. Create a file named `deno_net_bug.js` with the following content:
```js
import assert from 'node:assert/strict';
import { AsyncLocalStorage } from 'node:async_hooks';
import net from 'node:net';
import test from 'node:test';

// Run with:
//   deno test --allow-net playground/deno_net_bug.js
// or:
//   node --test playground/deno_net_bug.js

async function observeNetCallbacks() {
  const als = new AsyncLocalStorage();
  const observed = {
    serverConnection: null,
    clientConnect: null,
  };

  await new Promise((resolve, reject) => {
    als.run('server-context', () => {
      const server = net.createServer((socket) => {
        observed.serverConnection = als.getStore() ?? null;
        socket.end('ok');
        server.close();
      });

      server.once('error', reject);
      server.listen(() => {
        const address = server.address();
        const port = typeof address === 'object' && address ? address.port : address;

        als.run('client-context', () => {
          const client = net.connect(String(port), () => {
            observed.clientConnect = als.getStore() ?? null;
          });

          client.once('error', reject);
          client.once('data', () => {
            client.destroy();
            resolve();
          });
        });
      });
    });
  });

  return observed;
}

test('AsyncLocalStorage is preserved across node:net callbacks', async (t) => {
  const observed = await observeNetCallbacks();

  t.diagnostic(JSON.stringify(observed, null, 2));

  assert.deepEqual(observed, {
    serverConnection: 'server-context',
    clientConnect: 'client-context',
  });
});
```
2. Run with:
```bash
deno test --allow-net ./deno_net_bug.js
```

### Expected Behavior
Context should be preserved; test should pass.

### Actual Behavior
Test fails. Full output:
```
running 1 test from ./deno_net_bug.js
AsyncLocalStorage is preserved across node:net callbacks ...
------- output -------
DIAGNOSTIC: {
  "serverConnection": null,
  "clientConnect": null
}
----- output end -----
AsyncLocalStorage is preserved across node:net callbacks ... FAILED (5ms)

 ERRORS 

AsyncLocalStorage is preserved across node:net callbacks => ./deno_net_bug.js:49:1
error: AssertionError: Expected values to be strictly deep-equal:
+ actual - expected

  {
+   clientConnect: null,
+   serverConnection: null
-   clientConnect: 'client-context',
-   serverConnection: 'server-context'
  }

  assert.deepEqual(observed, {
         ^
    at NodeTestContext.<anonymous> (file:///.../deno_net_bug.js:54:10)

 FAILURES 

AsyncLocalStorage is preserved across node:net callbacks => ./playground/deno_net_bug.js:49:1

FAILED | 0 passed | 1 failed (10ms)

error: Test failed
```

### Additional Context
Same test passes in Nodejs and bun.

--- a/ext/node/polyfills/net.ts
+++ b/ext/node/polyfills/net.ts
@@ -182,6 +182,30 @@ let debug = debuglog("net", (fn) => {
 const kLastWriteQueueSize = Symbol("lastWriteQueueSize");
 const kBytesRead = Symbol("kBytesRead");
 const kBytesWritten = Symbol("kBytesWritten");
+// Holds the async context (AsyncLocalStorage / async_hooks) snapshot captured
+// when a connect request or listening handle is set up. The native libuv
+// callbacks in tcp_wrap/pipe_wrap invoke the completion callbacks directly,
+// outside of any microtask or nextTick, so the snapshot has to be captured at
+// registration time and restored before dispatching. Otherwise
+// `AsyncLocalStorage.getStore()` returns `undefined` inside `connect` and
+// `connection` handlers. See https://github.com/denoland/deno/issues/35154.
+const kAsyncContext = Symbol("kAsyncContext");
+
+// Runs `run` with the async context snapshot that was active when the
+// operation was initiated, restoring the previous context afterwards.
+function _runInAsyncContext(snapshot: any, run: () => void) {
+  if (snapshot === undefined) {
+    run();
+    return;
+  }
+  const prior = core.getAsyncContext();
+  core.setAsyncContext(snapshot);
+  try {
+    run();
+  } finally {
+    core.setAsyncContext(prior);
+  }
+}
 
 const DEFAULT_IPV4_ADDR = "0.0.0.0";
 const DEFAULT_IPV6_ADDR = "::";
@@ -460,6 +484,19 @@ function _normalizeArgs(args: unknown[]): NormalizedArgs {
 }
 
 function _afterConnect(
+  status: number,
+  handle: any,
+  req: PipeConnectWrap | TCPConnectWrap,
+  readable: boolean,
+  writable: boolean,
+) {
+  _runInAsyncContext(
+    handle?.[kAsyncContext],
+    () => _afterConnectImpl(status, handle, req, readable, writable),
+  );
+}
+
+function _afterConnectImpl(
   status: number,
   handle: any,
   req: PipeConnectWrap | TCPConnectWrap,
@@ -1595,6 +1632,11 @@ Socket.prototype.connect = function (...args) {
     _initSocketHandle(this);
   }
 
+  // Capture the async context now, while we are still running synchronously
+  // inside the caller's context. The DNS lookup that precedes the actual
+  // connect is async and would otherwise drop it before `_afterConnect` runs.
+  this._handle[kAsyncContext] = core.getAsyncContext();
+
   if (cb !== null) {
     this.once("connect", cb);
   }
@@ -2546,6 +2588,15 @@ function _emitListeningNT(server: Server) {
 function _onconnection(this: any, err: number, clientHandle?: Handle) {
   // deno-lint-ignore no-this-alias
   const handle = this;
+  _runInAsyncContext(
+    handle[kAsyncContext],
+    () => FunctionPrototypeCall(_onconnectionImpl, handle, err, clientHandle),
+  );
+}
+
+function _onconnectionImpl(this: any, err: number, clientHandle?: Handle) {
+  // deno-lint-ignore no-this-alias
+  const handle = this;
   const self = handle[ownerSymbol];
 
   debug("onconnection");
@@ -2660,6 +2711,7 @@ function _setupListenHandle(
 
   this[asyncIdSymbol] = _getNewAsyncId(this._handle);
   this._handle.onconnection = _onconnection;
+  this._handle[kAsyncContext] = core.getAsyncContext();
   this._handle[ownerSymbol] = this;
 
   // For TCP and Pipe handles, wrap the onconnection callback to create

--- a/tests/unit_node/async_hooks_test.ts
+++ b/tests/unit_node/async_hooks_test.ts
@@ -258,3 +258,50 @@ Deno.test(async function asyncLocalStoragePreservedInStreamFinished() {
 
   await promise;
 });
+
+// Regression test for https://github.com/denoland/deno/issues/35154. The
+// native libuv callbacks for node:net connect/accept invoke their completion
+// callbacks directly, so the async context that was active when the operation
+// was started has to be captured and restored, otherwise it is lost inside the
+// `connect` and `connection` handlers.
+Deno.test(async function asyncLocalStoragePreservedInNetCallbacks() {
+  const net = await import("node:net");
+  const als = new AsyncLocalStorage();
+  const observed: { serverConnection: unknown; clientConnect: unknown } = {
+    serverConnection: null,
+    clientConnect: null,
+  };
+
+  await new Promise<void>((resolve, reject) => {
+    als.run("server-context", () => {
+      const server = net.createServer((socket) => {
+        observed.serverConnection = als.getStore() ?? null;
+        socket.end("ok");
+        server.close();
+      });
+
+      server.once("error", reject);
+      server.listen(0, () => {
+        const addr = server.address()!;
+        const port = typeof addr === "string" ? addr : addr.port;
+
+        als.run("client-context", () => {
+          const client = net.connect(port, () => {
+            observed.clientConnect = als.getStore() ?? null;
+          });
+
+          client.once("error", reject);
+          client.once("data", () => {
+            client.destroy();
+            resolve();
+          });
+        });
+      });
+    });
+  });
+
+  assertEquals(observed, {
+    serverConnection: "server-context",
+    clientConnect: "client-context",
+  });
+});

--- a/tests/unit_node/async_hooks_test.ts
+++ b/tests/unit_node/async_hooks_test.ts
@@ -282,11 +282,10 @@ Deno.test(async function asyncLocalStoragePreservedInNetCallbacks() {
 
       server.once("error", reject);
       server.listen(0, () => {
-        const addr = server.address()!;
-        const port = typeof addr === "string" ? addr : addr.port;
+        const { port } = server.address() as { port: number };
 
         als.run("client-context", () => {
-          const client = net.connect(port, () => {
+          const client = net.connect({ port }, () => {
             observed.clientConnect = als.getStore() ?? null;
           });
 

